src/algebra/char_p.lean

```diff
@@ -313,3 +322,38 @@ by { intros h, apply @zero_ne_one R, symmetry, rw [←nat.cast_one, ring_char.sp
 end char_one
 
 end char_p
+
+section
+
+variables (n : ℕ) (R : Type*) [comm_ring R] [fintype R]
+
+lemma char_p_of_ne_zero (hn : fintype.card R = n) (hR : ∀ i < n, (i : R) = 0 → i = 0) :
+  char_p R n :=
```

Comment:
Not sure about this statement: there doesn't seem to be a `n ≠ 0` in the assumptions, and `hn` seems to be better suited to being proved at the use-site rather than the declaration-site.

Here is the most radically different suggestion I could think of, feel free to keep some parts the same though:
```suggestion
lemma char_p_card_of_coe_fin_injective (hR : ∀ (i : fin fintype.card R), (i : R) = 0 → i = 0) :
  char_p R (fintype.card R) :=
```

Reply:
I'm sorry, but what is the downside of my version? I think that very often you want to show that `R` has characteristic `n` for a preexisting given `n`. So I like the added flexibility of `hn`. And you can get pretty close to your version by just putting `rfl` into that argument slot.

Reply:
Right, we lose flexibility without `hn`, but don't we gain most of it back with `convert`? The reason I'm not convinced by this style is that it opens up a lot of design complexity: which definitions and which arguments should be a variable that is `eq`, and which should be an expression? So less arguments sounds better to me.

Of course, this is mostly hypothetical. If you expect that most invocations will be using `convert`, then I'm happy to let `hn` stay.

Reply:
I do think that in this case, Johan's phrasing is more convenient. It's likely very rare that you literally want `char_p` of a `fintype.card`, and also rare that `fintype.card` will be defeq to what you want.

Reply:
OK, let's keep `hn` :)